src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java

```diff
@@ -0,0 +1,134 @@
+package edu.kit.compiler.optimizations;
+
+import edu.kit.compiler.data.ast_nodes.ProgramNode;
+import edu.kit.compiler.lexer.Lexer;
+import edu.kit.compiler.lexer.StringTable;
+import edu.kit.compiler.logger.Logger;
+import edu.kit.compiler.parser.Parser;
+import edu.kit.compiler.semantic.DetailedNameTypeAstVisitor;
+import edu.kit.compiler.semantic.ErrorHandler;
+import edu.kit.compiler.semantic.NamespaceGatheringVisitor;
+import edu.kit.compiler.semantic.NamespaceMapper;
+import edu.kit.compiler.transform.IRVisitor;
+import edu.kit.compiler.transform.JFirmSingleton;
+import firm.Dump;
+import firm.Graph;
+import firm.Program;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class FirmGraphGenerator {
+
+    private static final ClassLoader systemClassLoader = ClassLoader.getSystemClassLoader();
+    private static final File tempDir;
+
+    // temp test resources
+    static {
+        String p = Objects.requireNonNull(systemClassLoader.getResource(".")).getFile();
+        tempDir = new File(p + "/tmp");
+        if (!tempDir.exists()) {
+            try {
+                Files.createDirectory(tempDir.toPath());
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        if (tempDir.exists()) {
+            // rm tmp
+            String[] entries = tempDir.list();
+            assert entries != null;
+            for (String s : entries) {
+                File currentFile = new File(tempDir.getPath(), s);
+                boolean success = currentFile.delete();
+                if (!success) {
+                    try {
+                        throw new RuntimeException("could not delete tmp file " + currentFile.getCanonicalPath());
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            }
+        }
+    }
+
+    public static void dumpGraph(Graph g) throws IOException {
+        JFirmSingleton.initializeFirmLinux();
```

Comment:
You have to assure that this is not called more than once.

Reply:
Actually, this is already the job of the singleton.

Reply:
Ah, I see.